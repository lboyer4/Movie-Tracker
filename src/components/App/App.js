import React, { Component } from 'react';
import './_App.scss';
import '../../index.scss';
import { Route } from 'react-router-dom';
import { apikey } from '../../apikey';
import { fetchMovie } from '../../utils/fetchMovie.js';
import { addMovies } from '../../actions'
import { connect } from 'react-redux';
import MovieContainer from '../../containers/MovieContainer/MovieContainer';
import Header from '../Header/Header';
import Login from '../../containers/Login/Login';
import Signup from '../../containers/Signup/Signup';
import FavoriteContainer from '../../containers/FavoriteContainer/FavoriteContainer';
import MovieDetails from '../MovieDetails/MovieDetails';
import { cleanMovies } from '../../utils/cleaners';

export class App extends Component {
  constructor() {
    super();
    this.state={
      movies: []
    }
  }

  componentDidMount = () => {
    const url = `https://api.themoviedb.org/3/movie/upcoming?api_key=${apikey}&language=en-US&page=1`;
   fetchMovie(url)
    .then(results => this.makeMovies(results.results)
    ) 
  }

  makeMovies = (movies) => {
    movies = cleanMovies(movies);
    this.props.addMovies(movies);
  }

  render() {
    return (
      <div className='App'>
        <Route exact path='/favorites' component= { FavoriteContainer } />
        <Route exact path='/' component= { Header  } />     
        <Route exact path='/' component= { MovieContainer }/>
        <Route exact path='/login' component={ Login } />
        <Route exact path='/signup' component={ Signup } />
        <Route path='/movies/:id' render={({ match }) => {
          const selectedMovie = this.props.movies.find(movie => {
            return movie.movie_id === parseInt(match.params.id)
           }) 
          if (selectedMovie) {         
            return <MovieDetails {...selectedMovie} />
            }
        }}
        />
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  movies: state.movies
})

export const mapDispatchToProps = (dispatch) => ({
  addMovies: (movies) => dispatch(addMovies(movies))
})

export default connect(mapStateToProps, mapDispatchToProps)(App)

