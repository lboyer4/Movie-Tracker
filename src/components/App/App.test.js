import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { mockUncleanMovie } from '../../utils/mockdata';
import {cleanMovies} from '../../utils/cleaners';
import { addMovies } from '../../actions';
import { fetchMovie } from '../../utils/fetchMovie.js';
import { apikey } from '../../apikey';

jest.mock('../../utils/cleaners');

jest.mock('../../utils/fetchMovie.js')
fetchMovie.mockImplementation(() => 
	Promise.resolve( {ok: true}))

const mockCleanMovies = jest.fn();
const mockMakeMovies = jest.fn();

describe('App', () => {
	let wrapper;

	beforeEach(() => {
		wrapper = shallow(
				<App 
				addMovies= {jest.fn()} 
				/>
		);
	});

	it('should match the snapshot with all data passed in', () => {
		expect(wrapper).toMatchSnapshot();
	});

	describe('componentDidMount', () => {

		it('should call fetch movie when mounted', () => {
			const wrapper = shallow(<App />)

			const mockUrl = `https://api.themoviedb.org/3/movie/upcoming?api_key=${apikey}&language=en-US&page=1`
			
			wrapper.instance().componentDidMount();

			expect(fetchMovie).toHaveBeenCalledWith(mockUrl)
		});
	});

	describe('makeMovies', () => {

		it('should call clean movies', () => {
			wrapper.instance().makeMovies([mockUncleanMovie])
			expect(cleanMovies).toHaveBeenCalled();
		});
	});

	describe('mapStateToProps', () => {

		it('should return an object with movies array', () => {

			const mockState = {
				movies: ['movies']
			}

			const expected = {
				movies: mockState.movies 
			}

			const mappedProps = mapStateToProps(mockState)
			expect(mappedProps).toEqual(expected)
		})
	})

	describe('mapDispatchToProps', () => {

		it('should call dispatch when using a function from MDTP', () => {

			const mockDispatch = jest.fn();
			const actionToDispatch = addMovies(['movies']);
			const mappedProps= mapDispatchToProps(mockDispatch)

			mappedProps.addMovies(['movies'])

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
		})
	})
})

