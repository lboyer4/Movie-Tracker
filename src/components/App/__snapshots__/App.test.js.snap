// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all data passed in 1`] = `
<div
  className="App"
>
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(FavoriteContainer)",
        "type": [Function],
      }
    }
    exact={true}
    path="/favorites"
  />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(Header)",
        "type": [Function],
      }
    }
    exact={true}
    path="/"
  />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(MovieContainer)",
        "type": [Function],
      }
    }
    exact={true}
    path="/"
  />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(Login)",
        "type": [Function],
      }
    }
    exact={true}
    path="/login"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/signup"
  />
  <Route
    path="/movies/:id"
    render={[Function]}
  />
</div>
`;
